File: Data Structures & Algorithms/balanced-binary-tree/submission-6.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        if(!root) return true;
        function dfs(root){
            if (!root) return 0;
            let leftDepth = dfs(root.left)
            if (leftDepth === -1) return -1;
            let rightDepth = dfs(root.right)
            if (rightDepth === -1) return -1;
            if((Math.max(leftDepth, rightDepth) - Math.min(leftDepth, rightDepth)) > 1) return -1;
            return 1 + Math.max(leftDepth, rightDepth);
        }
        if(dfs(root) === -1) return false;
        return true;
    }
}

File: Data Structures & Algorithms/binary-tree-longest-consecutive-sequence/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxLen = 0;
    int dfs(TreeNode* root, int curVal, int seqLen){
        if (!root) return seqLen;
        maxLen = max(maxLen, seqLen);
        if (root->val == curVal+1) {
            return max(maxLen, max(dfs(root->left, root->val, seqLen+1), dfs(root->right, root->val, seqLen+1)));
        } else {
            return max(maxLen, max(dfs(root->left, root->val, 1), dfs(root->right, root->val, 1)));
        }
    }

    int longestConsecutive(TreeNode* root) {
        return dfs(root, root->val-1, 0);
    }
};
